nano/node/election_occupancy.cpp

```diff
@@ -0,0 +1,53 @@
+#include <nano/lib/locks.hpp>
+#include <nano/lib/stats.hpp>
+#include <nano/node/active_transactions.hpp>
+#include <nano/node/election_occupancy.hpp>
+
+#include <boost/format.hpp>
+
+nano::election_occupancy::election_occupancy (nano::active_transactions & active, nano::stats & stats, size_t limit, nano::stat::detail detail) :
+	active{ active },
+	stats{ stats },
+	limit_m{ limit },
+	detail{ detail }
+{
+}
+
+void nano::election_occupancy::clean () const
+{
+	auto initial_size = elections.size ();
+	elections.remove_if ([this] (std::weak_ptr<nano::election> & election) {
+		std::shared_ptr<nano::election> election_l;
+		auto result = (election_l = election.lock ()) == nullptr || !active.active (election_l->qualified_root);
+		return result;
+	});
+}
+
+size_t nano::election_occupancy::limit () const
+{
+	return limit_m;
+}
+
+bool nano::election_occupancy::available () const
+{
+	clean ();
+	auto result = elections.size () < limit ();
+	return result;
+}
+
+nano::election_insertion_result nano::election_occupancy::insert (std::shared_ptr<nano::block> const & block)
+{
+	if (!available ())
+	{
+		return { nullptr, false };
+	}
+	nano::unique_lock<nano::mutex> lock{ active.mutex };
```

Comment:
Locking mutexes of outside classes is something I went out of my way to eliminate. It's a very bad pattern, makes reasoning about multithreaded classes unnecessarily difficult.